### PR TITLE
Support revoke for sdh

### PR DESCRIPTION
**Pull request by mlsteele**

### CODE REVIEW

File: go/engine/revoke.go [hunk 1]
```diff
@@ -122,37 +126,104 @@ func (e *RevokeEngine) Run(ctx *Context) error {
...
+	sigKey, encKey, err := e.getDeviceSecretKeys(ctx, me)
 	if err != nil {
 		return err
 	}
-	if err = sigKey.CheckSecretKey(); err != nil {
-		return err
+
+	if e.G().Env.GetEnableSharedDH() {
+		if encKey == nil {
+			return fmt.Errorf("Missing encryption key")
```
Comment (maxtaco) [inline go/engine/revoke.go]: `errors.New`

File: go/engine/revoke.go [hunk 2]
```diff
@@ -122,37 +126,104 @@ func (e *RevokeEngine) Run(ctx *Context) error {
...
-	}
-	sigKey, err := e.G().Keyrings.GetSecretKeyWithPrompt(ctx.SecretKeyPromptArg(ska, "to revoke another key"))
+	sigKey, encKey, err := e.getDeviceSecretKeys(ctx, me)
 	if err != nil {
 		return err
 	}
-	if err = sigKey.CheckSecretKey(); err != nil {
-		return err
+
+	if e.G().Env.GetEnableSharedDH() {
```
Comment (maxtaco) [inline go/engine/revoke.go]: Can you factor this change out into a helper routine?

File: go/engine/revoke.go [hunk 3]
```diff
@@ -122,37 +126,104 @@ func (e *RevokeEngine) Run(ctx *Context) error {
...
+	addingNewSDH := false
+
+	// Only used if addingNewSDH
+	var newSdhGeneration keybase1.SharedDHKeyGeneration
+	var sdhGen *libkb.NaclKeyGen
+	var newSdhKeyPair libkb.NaclDHKeyPair
+
+	if e.G().Env.GetEnableSharedDH() {
+		// Sync the SDH keyring
+		err = e.G().BumpSharedDHKeyring()
```
Comment (maxtaco) [inline go/engine/revoke.go]: comment doesn't match the code.

File: go/engine/revoke.go (hunk 2 above)
Comment (maxtaco) [inline go/engine/revoke.go]: We're going to need to run this code from outside this engine when a user is upgraded from to using DH from legacy.

File: go/engine/revoke.go [hunk 4]
```diff
@@ -162,6 +233,154 @@ func (e *RevokeEngine) Run(ctx *Context) error {
...
+		encKey = &encKey2
+		if err = encKey.CheckSecretKey(); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return sigKey, encKey, err
+}
+
+func (e *RevokeEngine) makeSdhSig(ctx *Context, me *libkb.User, sdhGen *libkb.NaclKeyGen,
```
Comment (maxtaco) [inline go/engine/revoke.go]: `sharedDH` instead of `Sdh`

File: go/engine/revoke_test.go [hunk 5]
```diff
@@ -119,6 +140,17 @@ func TestRevokePaperDevice(t *testing.T) {
 	require.NoError(t, err)
 
 	assertNumDevicesAndKeys(tc, u, 1, 2)
+
+	checkSharedDHKeyring(t, tc.G)
+}
+
+func checkSharedDHKeyring(t *testing.T, g *libkb.GlobalContext) {
+	// double check that the sdh keyring is correct
```
Comment (maxtaco) [inline go/engine/revoke_test.go]: Does this check tell that the generation on the server was actually bumped/

Comment (maxtaco) [inline go/engine/revoke_test.go]: Does this check tell that the generation on the server was actually bumped?

**Comment by mlsteele** [comment]
Enacted some feedback and added some more tests. This won't pass CI until after updating to do chaining.

**Comment by mlsteele** [comment]
new one coming soon

**Pull request closed without merging**